Provide automation subject fields as tags in email block editor

lysyjan opened the pull request:
## Description

_N/A_

## Code review notes

_N/A_

## QA notes

1. Install WooCommerce, WooCommerce Bookings, and WooCommerce Subscriptions plugins
2. Set `Use block email editor for automation emails` to `Yes`
3. Create automation, for example, `Purchased a product`
4. Design an email in the block email editor and try to use WooCommerce Personalization tags
5. Verify that tags are replaced in the browser preview, email preview, and sent emails

Repeat steps 3 - 5 for the different automation triggers for WooCommerce Bookings and Subscriptions.
There should be related personalization tags based on the automation subjects.

## Linked PRs

https://github.com/mailpoet/mailpoet-premium/pull/991

## Linked tickets

STOMAIL-7353

## After-merge notes

_N/A_

## Tasks

- [x] I have added a changelog entry (`./do changelog:add --type=<type> --description=<description>`)
- [ ] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [x] I added sufficient test coverage
- [ ] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes

## Preview

[Preview in WordPress Playground](https://account.mailpoet.com/playground/new/branch:stomail-7353-provide-subject-fields-as-tags-in-new-editor)

_The latest successful build from `stomail-7353-provide-subject-fields-as-tags-in-new-editor` will be used. If none is available, the link won't work._

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Automation subject fields are now available as personalization tags in the email block editor, previews, and sends.
  * Editor previews and preview emails include sample WooCommerce order and customer data so personalization tags render correctly.
* **Documentation**
  * Added a changelog entry describing these personalization and preview enhancements.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

lysyjan committed:
Extend personalization tag support for automation emails with WooCommerce data

Enhancements include:
- Dynamically extending WooCommerce tags for automation emails.
- Adding sample order/customer data for preview generation in automation emails.
- Improving context handling for personalization, leveraging WooCommerce dummy data when previewing automation emails.

lysyjan committed:
Extend personalization tags and context handling for automation emails

Enhancements include:
- Refactoring WooCommerce tag extensions to use automation subjects.
- Adding filters for extending tags and sample data for previews.
- Improving personalization context with extensibility for automation subjects.
Release 5.17.0

lysyjan committed:
Refactor automation subject handling for personalized emails

Enhancements include:
- Moving subject storage to `AutomationRunStorage` for improved flexibility.
- Dynamic loading of WooCommerce order and customer data from automation subjects.
- Extending personalization tags with additional filters and actions.

lysyjan committed:
Add missing method stubs to PHPStan custom stubs for improved type compatibility

lysyjan committed:
Add changelog for personalization tags by automation subjects

lysyjan committed:
Add `get_time` method stub to PHPStan custom stubs for type compatibility

lysyjan committed:
Improve `PersonalizationTagManager` to parse `post` ID using `parse_url` and `parse_str` for safer and more reliable URL handling

NeosinneR commented on mailpoet/lib/Cron/Workers/SendingQueue/Tasks/Newsletter.php:
```diff
@@ -289,11 +294,69 @@ public function prepareNewsletterForSending(NewsletterEntity $newsletter, Subscr
+
+      $queueMeta = $queue->getMeta();
+      if (
+        is_array($queueMeta)
```
Just a nitpick, but isn't `is_array()` redundant? `isset()` should return false, even if `$queueMeta` is `null` or any other type.

NeosinneR commented on mailpoet/lib/EmailEditor/Integrations/MailPoet/PersonalizationTagManager.php:
```diff
@@ -16,22 +20,96 @@ class PersonalizationTagManager {
+            $postId = (int)$params['post'];
+          }
+        }
+      }
```
@lysyjan I see 👎, so I take it you acknowledge this. I wanted to ask the same as CodeRabbit did, that relying on a referer feels fragile as I'm somewhat certain that you can somehow set the browser to not send referers, even if fot the same-site requests. 

Any chance of adding the `post_id` parameter directly to email editor when making the request for personalization tags? 

This can be a temporary workaround, as I'm aware releasing and then updating email editor is a non-trivial task, but long-term, I think we shouldn't rely on referer to be present.

NeosinneR commented on mailpoet/lib/Newsletter/Preview/SendPreviewController.php:
```diff
@@ -103,4 +113,119 @@ public function sendPreview(NewsletterEntity $newsletter, string $emailAddress)
+      return $dummyOrder instanceof \WC_Order ? $dummyOrder : null;
+    } catch (\Throwable $e) {
+      // If reflection fails, try the filter as fallback
+      $dummyOrder = $this->wp->applyFilters('woocommerce_email_preview_dummy_order', null, null);
```
What's the expectation here? Applying the same filter as WooCommerce core applies, with `null` will just return `null` no?

NeosinneR commented on mailpoet/lib/Newsletter/Preview/SendPreviewController.php:
```diff
@@ -103,4 +113,119 @@ public function sendPreview(NewsletterEntity $newsletter, string $emailAddress)
+
+      $dummyOrder = $method->invoke($emailPreview);
+
+      return $dummyOrder instanceof \WC_Order ? $dummyOrder : null;
```
Would it make sense to make the method public in WC Core?

NeosinneR commented on mailpoet/lib/EmailEditor/Integrations/MailPoet/PersonalizationTagManager.php:
```diff
@@ -16,22 +20,96 @@ class PersonalizationTagManager {
+      $postId = null;
+      if (isset($_SERVER['HTTP_REFERER'])) {
+        $referer = sanitize_text_field(wp_unslash($_SERVER['HTTP_REFERER']));
+        $queryString = parse_url($referer, PHP_URL_QUERY);
```
I'd suggest using `$this->wp->wpParseUrl()` which is a [wrapper for `parse_url()`](https://developer.wordpress.org/reference/functions/wp_parse_url/) that handles some inconsistencies and edge cases.

NeosinneR reviewed (commented):
Nice work, @lysyjan. This is great addition to automations. I reviewed the code and briefly tested the functionality, and have a couple of notes below. No major issues.

lysyjan committed:
Address code review feedback

Changes include:
- Rename unused `$server` parameter to `$_server` in PersonalizationTagManager to satisfy static analysis
- Remove redundant `is_array()` check in Newsletter.php since `getMeta()` returns `?array` and `isset()` handles null gracefully
- Remove ineffective fallback filter for dummy order that would always return null
- Add backward compatibility for `get_dummy_order()` visibility: call directly if public
- Use `wpParseUrl()` wrapper instead of native `parse_url()

NeosinneR reviewed (approved):
Thanks for the updates. The functionality and code looks good, moving to QA review.

lysyjan committed:
Simplify personalization tags filtering using action hook

- Use woocommerce_email_editor_personalization_tags_for_post action
- Add mailpoet_personalization_tags_category_subjects_mapping filter
- Remove REST dispatch hooks in favor of action-based approach
- Add post_id to preloaded personalization tags URL

lysyjan committed:
Update automation email handling to use post meta flag

Replaced automation email detection logic with a meta flag set during post creation.
This ensures more reliable identification of automation emails and prevents duplication
issues caused by other plugins.

lysyjan committed:
Prevent extending unknown personalization tag categories in `PersonalizationTagManager`

lysyjan commented on mailpoet/lib/EmailEditor/Integrations/MailPoet/PersonalizationTagManager.php:
```diff
@@ -147,4 +199,123 @@ public function convertLinksToShortcodes(array $emailContent): array {
+    foreach ($tags as $tag) {
+      $postTypes = $tag->get_post_types();
+
+      // If this is a WooCommerce tag (Order, Customer, Site, Store) and doesn't already support mailpoet_email
+      if (!empty($postTypes) && !in_array(EmailEditor::MAILPOET_EMAIL_POST_TYPE, $postTypes, true)) {
```
I think we can skip it because it's a detail.

costasovo commented on mailpoet/lib/Newsletter/Preview/SendPreviewController.php:
```diff
@@ -103,4 +121,126 @@ public function sendPreview(NewsletterEntity $newsletter, string $emailAddress)
+    }
+
+    try {
+      $emailPreview = \Automattic\WooCommerce\Internal\Admin\EmailPreview\EmailPreview::instance();
```
We shouldn't use classes from `\Automattic\WooCommerce\Internal` namespace in other plugins. Could we create the order here or perhaps make a change in Woo to expose `get_dummy_order` somehow from public namespace?

costasovo reviewed (commented):
I did just a brief walkthrough and will test it after the changes in the editor package are merged. 
I noticed the use of an internal Woo class. We should avoid this. Please see my other comment.

lysyjan committed:
Add `WooCommerceDummyData` integration for preview sample data

- Replaced legacy dummy order and customer generation logic with `WooCommerceDummyData` service.
- Updated `SendPreviewController` to include `WooCommerceDummyData` dependency.
- Simplified sample data addition for order and customer previews.

lysyjan committed:
Add `set_persons` method stub to PHPStan custom stubs for type compatibility

costasovo reviewed (approved):
@lysyjan Thank you for the updates!

I confirm I was able to use Woo tags in an automation email for an order and the tags were correctly rendered in preview and sent email.

<img width="630" height="702" alt="Image" src="https://github.com/user-attachments/assets/5fe7456c-d087-4a7a-9180-8f85c27d0362" />
<img width="608" height="643" alt="Image" src="https://github.com/user-attachments/assets/3a3835fc-6a8e-4709-8996-bdb5db256882" />
